File: tuning_fun_fitting.m

% show the tuning function 
kappa = 3; 
Amp = 30; 
B = 10; 
tf = @(x) Amp*exp(kappa*cos(x)-kappa) + B; 

% simulate data
n = 12; 
x = linspace(-pi,pi-2*pi/n,n);
y = tf(x) + randn(size(x));

% fit data with least squares
fn = @(x, P) P(1)*exp(P(2)*cos(x)-P(2)) + P(3); 
obj_fn = @(P) sum((fn(x, P) - y).^2,'all');
p0 = rand(1,3)*3; 
P = fminunc(obj_fn, p0)

clf
nexttile
hold on 
xx = linspace(-pi,pi,1e3);
plot(xx, tf(xx))
scatter(x, y, 'filled')
plot(xx, fn(xx, P))


legend({'Ground truth','Simulated data','Fit'})
xlabel('Orientation (2x rad.)')
ylabel('Firing rate')
title('Tuning function')

File: tuning_fun_fitting.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.optimize import minimize

# show the tuning function
kappa = 3
Amp = 30
B = 10
tf = lambda x: Amp * np.exp(kappa * np.cos(x) - kappa) + B

# simulate data
n = 12
x = np.linspace(-np.pi, np.pi - 2*np.pi/n, n)
y = tf(x) + np.random.randn(*x.shape)

# fit data with least squares
fn = lambda x, P: P[0] * np.exp(P[1] * np.cos(x) - P[1]) + P[2]
obj_fn = lambda P: np.sum((fn(x, P) - y) ** 2)
p0 = np.random.rand(3) * 3
res = minimize(obj_fn, p0)
P = res.x
print(P)

plt.figure(figsize=(8, 4))

xx = np.linspace(-np.pi, np.pi, int(1e3))
plt.plot(xx, tf(xx), 'k--')
plt.scatter(x, y)
plt.plot(xx, fn(xx, P))

plt.legend(["Ground truth", "Simulated data", "Fit"])
plt.xlabel("Orientation (2x rad.)")
plt.ylabel("Firing rate")
plt.title("Tuning function")

plt.tight_layout()
plt.show()